torch/_dynamo/optimizations/analysis.py

```diff
@@ -118,35 +118,14 @@ def run(self, *args):
             self.env.clear()
 
 
-def has_mutation(gm, example_inputs, inputs_only=False):
+def has_mutation(gm, example_inputs, fake_mode, inputs_only=False):
     """Check if the graph module has any form of mutation.  If inputs_only is
     true, we only check for mutation of inputs"""
     # TODO - moco gives bad accuracy with Aliasing. gm is getting mutated in a bad way.
-
     if fake_tensors_available and config.fake_tensor_propagation:
-
-        def _wrap_to_fake_tensor(t, *, f_mode):
-            if type(t) in (torch.Tensor, torch.nn.Parameter):
-                static_shapes_ = config.dynamic_shapes is False
-                return fake_mode.from_tensor(
-                    t, static_shapes=config.dynamic_shapes is not False
-                )
-            else:
-                return t
-
-        # Our analysis pass should use dynamic shape tensor inputs
-        # when dynamic shapes are enabled.
-        # We don't actually care about the guards that are created
-        # on those shapes though, so just create a fresh ShapeEnv here.
-        from torch.fx.experimental.symbolic_shapes import ShapeEnv
-
-        fake_mode = FakeTensorMode(
-            shape_env=ShapeEnv() if config.dynamic_shapes else None
-        )
-        fake_wrapper = functools.partial(_wrap_to_fake_tensor, f_mode=fake_mode)
-        example_inputs = tree_map(fake_wrapper, example_inputs)
+        assert fake_mode, "Fake mode must be passed in"
         new_gm = deepcopy_to_fake_tensor(gm, fake_mode)
-        with fake_mode, enable_python_dispatcher():
+        with enable_python_dispatcher():
```

Comment:
So `FakeTensorMode` is no longer enabled when this pass runs? The bad situation I'm imagining could happen is:

(1) The graph contains a factory function, so interpreting the graph ends up creating a real tensor with real sizes (bad because we do unnecessary compute, etc)

(2) If the graph contains a SymInt-related op, like `s1.__floordiv__(s2)`, and we grab the real sizes off of our real tensor, then we'll break on this pass when we run that op.

The original idea behind me adding that temporary extra `FakeTensorMode` and using it here was that:
(1) We only want to create fake tensors and symbolic shapes when running this pass
(2) Using the existing ShapeEnv in dynamo might be bad, because we could end up installing a bunch of redundant guards in it when we run this pass. It sounds like this was the wrong thing to do though, since mixing tensors across multiple FakeTensorModes causes issues.

Reply:
That was just my brain dump understanding. If we're going to try to remove this pass soon anyway though, and this change fixed existing issues in the mean time, then I'm for landing it

Reply:
We can keep fake_mode here, we need it for the deepcopy anyway, I just thought it was spurious.